costmap:
  costmap:
    ros__parameters:
      update_frequency: 10.0
      publish_frequency: 1.0
      global_frame: vodom
      robot_base_frame: base_link
      transform_tolerance: 1.2
      use_sim_time: False
      rolling_window: true
      width: 10
      height: 10
      resolution: 0.2
      robot_radius: 0.2
      plugins: ["obstacle_layer", "obstacle_layer_rgbd"] #", "inflation_layer"]
      obstacle_layer:
        plugin: "mcr_tracking_components::ObstacleLayer"
        enabled: True
        use_tracking_mode: True 
        footprint_clearing_enabled: True
        tracking_pose_invisible_radius: 0.8
        tracking_pose_topic: tracking_pose
        observation_sources: scan
        scan:
          topic: scan
          max_obstacle_height: 20.0
          min_obstacle_height: -20.0
          clearing: True
          marking: True
          data_type: "LaserScan"
          obstacle_max_range: 6.0
          obstacle_min_range: 0.3
          raytrace_min_range: 0.3
          raytrace_max_range: 6.0
      obstacle_layer_rgbd:
        plugin: "mcr_tracking_components::ObstacleLayer"
        enabled: True
        use_tracking_mode: True 
        footprint_clearing_enabled: True
        tracking_pose_invisible_radius: 0.8
        tracking_pose_topic: tracking_pose
        observation_sources: scan
        scan:
          topic: camera/depth/color/points1
          max_obstacle_height: 0.2
          min_obstacle_height: -0.05
          clearing: True
          marking: True
          data_type: "PointCloud2"
          obstacle_max_range: 6.0
          obstacle_min_range: 0.3
          raytrace_min_range: 0.3
          raytrace_max_range: 6.0
         
      static_layer:
        map_subscribe_transient_local: True
      always_send_full_costmap: True

tracking_base:
  ros__parameters:
    use_sim_time: False
    odom_topic: odom_out
    controller_frequency: 20.0
    global_frame: vodom
    base_frame: base_link
    goal_updater_topic: tracking_pose
    overtime: 1.50
    max_pose_inuse: 0
    dist_throttle: 0.3
    keep_distance: 1.3 
    # teb parameters
    teb:
      teb_autosize: 1.0
      dt_ref: 0.3
      dt_hysteresis: 0.1
      max_samples: 500
      global_plan_overwrite_orientation: True
      allow_init_with_backwards_motion: False
      max_global_plan_lookahead_dist: 3.0
      global_plan_viapoint_sep: 0.5
      global_plan_prune_distance: 1.0
      exact_arc_length: False
      feasibility_check_no_poses: 2
      publish_feedback: False

      # Robot
      max_vel_x: 1.5
      max_vel_x_backwards: 0.1
      max_vel_y: 0.3
      max_vel_theta: 1.6
      acc_lim_x: 1.5
      acc_lim_y: 0.2
      acc_lim_theta: 1.6
      min_turning_radius: 0.0

      footprint_model:
        type: "line"
        line_start: [-0.12, 0.0] # for type "line"
        line_end: [0.12, 0.0] # for type "line"        
      # GoalTolerance
      free_goal_vel: False

      # Obstacles

      min_obstacle_dist: 0.32
      inflation_dist: 0.2
      include_costmap_obstacles: True
      costmap_obstacles_behind_robot_dist: 1.0
      obstacle_poses_affected: 15

      dynamic_obstacle_inflation_dist: 0.6
      include_dynamic_obstacles: True 

      costmap_converter_plugin: "costmap_converter::CostmapToPolygonsDBSMCCH"
      # costmap_converter_plugin: "costmap_converter::CostmapToLinesDBSMCCH"
      costmap_converter_spin_thread: True
      costmap_converter_rate: 10

      # Optimization

      no_inner_iterations: 5
      no_outer_iterations: 4
      optimization_activate: True
      optimization_verbose: False
      penalty_epsilon: 0.1
      obstacle_cost_exponent: 4.0
      weight_max_vel_x: 1.0
      weight_max_vel_y: 0.5
      weight_max_vel_theta: 0.5
      weight_acc_lim_x: 0.5
      weight_acc_lim_y: 0.5
      weight_acc_lim_theta: 0.5
      weight_kinematics_nh: 4.0
      weight_kinematics_forward_drive: 1.0
      weight_kinematics_turning_radius: 0.0
      weight_optimaltime: 0.5 # must be > 0
      weight_shortest_path: 0.5
      weight_obstacle: 2000.0
      weight_inflation: 0.0
      weight_dynamic_obstacle: 10.0 # not in use yet
      weight_dynamic_obstacle_inflation: 0.2
      weight_viapoint: 0.1
      weight_adapt_factor: 2.0

      # Recovery

      shrink_horizon_backup: True
      shrink_horizon_min_duration: 10.0
      oscillation_recovery: True
      oscillation_v_eps: 0.1
      oscillation_omega_eps: 0.1
      oscillation_recovery_min_duration: 10.0
      oscillation_filter_duration: 10.0
